File: src/pocketcloud/cloud/util/FileUtils.php
<?php

namespace pocketcloud\cloud\util;

use pocketcloud\cloud\exception\ExceptionHandler;
use pocketcloud\cloud\terminal\log\CloudLogger;
use Throwable;

final class FileUtils {

    public static function copyFile(string $src, string $dst): bool {
        return ExceptionHandler::tryCatch(
            function (string $src, string $dst): bool {
                return copy($src, $dst);
            },
            "Failed to copy " . $src . " to " . $dst,
            null,
            $src, $dst
        );
    }

    public static function createDir(string $path): bool {
        return ExceptionHandler::tryCatch(
            function (string $path): bool {
                if (is_dir($path)) return true;
                $previousPath = substr($path, 0, strrpos($path, DIRECTORY_SEPARATOR, -2) + 1);
                $return = self::createDir($previousPath);
                return $return && is_writable($previousPath) && mkdir($path);
            },
            "Failed to create directory: " . $path,
            null,
            $path
        ) ?? false;
    }

    public static function filePutContents(string $filePath, string $content): int|false {
        return ExceptionHandler::tryCatch(
            function (string $filePath, string $content): int|false {
                return file_put_contents($filePath, $content);
            },
            "Failed to write in file: " . $filePath,
            null,
            $filePath, $content
        ) ?? false;
    }

    public static function fileGetContents(string $filePath): ?string {
        return ExceptionHandler::tryCatch(
            function (string $filePath): string {
                return file_get_contents($filePath);
            },
            "Failed to read file: " . $filePath,
            null,
            $filePath
        );
    }

    public static function copyDirectory(string $src, string $dst): bool {
        return ExceptionHandler::tryCatch(
            function (string $src, string $dst): bool {
                $src = rtrim($src, DIRECTORY_SEPARATOR);
                $dst = rtrim($dst, DIRECTORY_SEPARATOR);
                self::createDir($src);
                self::createDir($dst);

                foreach (array_diff(scandir($src), [".", ".."]) as $file) {
                    try {
                        if (filetype($src . DIRECTORY_SEPARATOR . $file) == "dir") {
                            self::copyDirectory($src . DIRECTORY_SEPARATOR . $file, $dst . DIRECTORY_SEPARATOR . $file);
                        } else {
                            try {
                                copy($src . DIRECTORY_SEPARATOR . $file, $dst . DIRECTORY_SEPARATOR . $file);
                            } catch (Throwable) {
                                CloudLogger::get()->debug("Can't copy file from: " . $src . DIRECTORY_SEPARATOR . $file . " to " . $dst . DIRECTORY_SEPARATOR . $file);
                            }
                        }
                    } catch (Throwable) {}
                }
                return false;
            },
            null,
            null,
            $src, $dst
        );
    }

    public static function unlinkFile(string $filePath): bool {
        return ExceptionHandler::tryCatch(
            fn() => @unlink($filePath),
            "Failed to unlink file: " . $filePath
        ) ?? false;
    }

    public static function removeDirectory(string $directoryPath): bool {
        return ExceptionHandler::tryCatch(
            function (string $directoryPath): bool {
                if (@is_dir($directoryPath)) {
                    foreach (array_diff(scandir($directoryPath), [".", ".."]) as $file) {
                        $filePath = rtrim($directoryPath, "/") . "/" . $file;
                        if (is_file($filePath)) self::unlinkFile($filePath);
                        else if (is_dir($filePath)) self::removeDirectory($filePath);
                    }
                    return rmdir($directoryPath);
                }

                return false;
            },
            "Failed to remove directory: " . $directoryPath,
            null,
            $directoryPath
        ) ?? false;
    }

    public static function jsonDecode(string $jsonString, int $depth = 512): ?array {
        return ExceptionHandler::tryCatch(
            function (string $jsonString, int $depth): ?array {
                $decode = json_decode($jsonString, true, $depth);
                return !$decode ? null : $decode;
            },
            "Failed to json decode: " . $jsonString,
            null,
            $jsonString, $depth
        );
    }

    public static function jsonEncode(array $jsonArray): ?string {
        return ExceptionHandler::tryCatch(
            function (array $jsonArray): ?string {
                $encode = json_encode($jsonArray);
                return !$encode ? null : $encode;
            },
            "Failed to json encode",
            null,
            $jsonArray
        );
    }
}

File: src/main/java/de/pocketcloud/cloud/util/FileUtils.java
package de.pocketcloud.cloud.util;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import org.yaml.snakeyaml.Yaml;

import java.lang.reflect.Type;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.Comparator;
import java.util.Map;
import java.util.Set;
import java.util.stream.Stream;

public final class FileUtils {

    public static final Gson PRETTY_GSON = new GsonBuilder().setPrettyPrinting().create();
    public static final Gson GSON = new Gson();
    public static final Yaml YAML = new Yaml();

    public static String extensionOf(Path path) {
        String fileName = path.getFileName().toString();
        int dotIndex = fileName.lastIndexOf(".");
        String extension = null;
        if (dotIndex > 0 && dotIndex < fileName.length() - 1) extension = fileName.substring(dotIndex + 1);
        return extension;
    }

    public static void createDirs(Path... path) {
        for (Path s : path) createDir(s);
    }

    public static boolean createDir(Path path) {
        try {
            Files.createDirectories(path);
            return true;
        } catch (Exception e) {
            throw new RuntimeException("Failed to create directory: " + path, e);
        }
    }

    public static boolean filePutContents(Path filePath, String content) {
        try {
            Files.createDirectories(filePath.getParent());
            Files.writeString(filePath, content);
            return true;
        } catch (Exception e) {
            throw new RuntimeException("Failed to write file: " + filePath, e);
        }
    }

    public static String fileGetContents(Path filePath) {
        return fileGetContents(filePath, "");
    }

    public static String fileGetContents(Path filePath, String defaultValue) {
        try {
            if (!Files.exists(filePath)) return defaultValue;
            return Files.readString(filePath);
        } catch (Exception e) {
            throw new RuntimeException("Failed to read file: " + filePath, e);
        }
    }

    public static boolean rename(Path from, Path to) {
        try {
            Files.move(from, to, StandardCopyOption.REPLACE_EXISTING);
            return true;
        } catch (Exception e) {
            throw new RuntimeException("Failed to rename " + from + " to " + to, e);
        }
    }

    public static boolean unlinkFile(Path filePath) {
        try {
            return Files.deleteIfExists(filePath);
        } catch (Exception e) {
            throw new RuntimeException("Failed to delete file: " + filePath, e);
        }
    }

    public static boolean copyDirectory(Path source, Path destination, Set<String> exclusions) {
        try {
            if (!Files.isDirectory(source)) {
                throw new IllegalArgumentException("Source directory does not exist: " + source);
            }

            try (Stream<Path> paths = Files.walk(source)) {
                paths.forEach(path -> {
                    try {
                        Path relative = source.relativize(path);
                        String name = relative.toString();

                        for (String ex : exclusions) {
                            if (name.startsWith(ex) || path.getFileName().toString().equals(ex)) {
                                return;
                            }
                        }

                        if (Files.isDirectory(path)) {
                            Files.createDirectories(destination);
                        } else {
                            Files.createDirectories(destination.getParent());
                            Files.copy(path, destination, StandardCopyOption.REPLACE_EXISTING);
                        }
                    } catch (Exception e) {
                        throw new RuntimeException("Failed copying: " + path, e);
                    }
                });
            }

            return true;
        } catch (Exception e) {
            throw new RuntimeException("Failed to copy directory: " + source + " -> " + destination, e);
        }
    }

    public static boolean removeDirectory(Path dir) {
        try {
            if (!Files.exists(dir)) return false;

            try (Stream<Path> paths = Files.walk(dir).sorted(Comparator.reverseOrder())) {
                paths.forEach(path -> {
                    try {
                        Files.deleteIfExists(path);
                    } catch (Exception e) {
                        throw new RuntimeException("Failed deleting: " + path, e);
                    }
                });
            }

            return true;
        } catch (Exception e) {
            throw new RuntimeException("Failed to remove directory: " + dir, e);
        }
    }

    public static String encodeJson(Object obj) {
        return GSON.toJson(obj);
    }

    public static boolean encodeJsonFile(Path filePath, Object obj) {
        return filePutContents(filePath, encodeJson(obj));
    }

    public static <T> T decodeJson(String json, Class<T> clazz) {
        return GSON.fromJson(json, clazz);
    }

    @SuppressWarnings("unchecked")
    public static <T> T decodeJson(String json, Type type) {
        return GSON.fromJson(json, type);
    }

    public static <T> T decodeJson(String json, TypeToken<T> typeToken) {
        return GSON.fromJson(json, typeToken);
    }

    public static <T> T decodeJsonFile(Path filePath, Class<T> clazz) {
        return decodeJson(fileGetContents(filePath), clazz);
    }

    public static <T> T decodeJsonFile(Path filePath, Type type) {
        return decodeJson(fileGetContents(filePath), type);
    }

    public static <T> T decodeJsonFile(Path filePath, TypeToken<T> typeToken) {
        return decodeJson(fileGetContents(filePath), typeToken);
    }

    public static String emitYaml(Object data) {
        return YAML.dump(data);
    }

    public static boolean emitYamlFile(Path filePath, Object data) {
        return filePutContents(filePath, emitYaml(data));
    }

    public static Map<String, Object> parseYaml(String yaml) {
        return YAML.load(yaml);
    }

    public static Map<String, Object> parseYamlFile(Path filePath) {
        return parseYaml(fileGetContents(filePath));
    }
}